samples/System.IO.Pipelines.Samples/HttpServer/HttpResponseStream.cs

```diff
@@ -0,0 +1,121 @@
+
+namespace System.IO.Pipelines.Samples.Http
+{
+    public class HttpResponseStream<TContext> : Stream
+    {
+        private readonly static Task<int> _initialCachedTask = Task.FromResult(0);
+        private Task<int> _cachedTask = _initialCachedTask;
+
+        private readonly HttpConnection<TContext> _connection;
+
+        public HttpResponseStream(HttpConnection<TContext> connection)
+        {
+            _connection = connection;
+        }
+
+        public override bool CanRead => false;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => true;
+
+        public override long Length
+        {
+            get
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        public override long Position
+        {
+            get
+            {
+                throw new NotSupportedException();
+            }
+            set
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            WriteAsync(buffer, offset, count, CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken token)
+        {
+            return _connection.WriteAsync(new Span<byte>(buffer, offset, count));
+        }
+
+        public override void Flush()
+        {
+            // No-op since writes are immediate.
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            // No-op since writes are immediate.
+            return Task.FromResult(0);
+        }
+
+#if NET451
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            var task = WriteAsync(buffer, offset, count, default(CancellationToken), state);
+            if (callback != null)
+            {
+                task.ContinueWith(t => callback.Invoke(t));
+            }
+            return task;
+        }
+
+        public override void EndWrite(IAsyncResult asyncResult)
+        {
+            ((Task<object>)asyncResult).GetAwaiter().GetResult();
+        }
+
+        private Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken, object state)
+        {
+            var tcs = new TaskCompletionSource<object>(state);
+            var task = WriteAsync(buffer, offset, count, cancellationToken);
```

Comment:
>var task = WriteAsync(buffer, offset, count, cancellationToken); [](start = 12, length = 64)

In cases like this, where the async operation is likely to complete synchronously, you're better off adding a fast path, e.g.
```C#
Task task = WriteAsync(...);
if (task.Status == TaskStatus.RanToCompletion)
{
    return task;
}

var tcs = ...
...
```